Implement BIP143 test vectors

tcharding opened the pull request:
Draft because I would like to tackle script tagging before this is considered for merging. The code in `tests/bip143.rs` now fully implements the test vectors in the BIP. There is one bug in the `SighashCache` that I've uncovered, see `FIXME` in `tests/bip143.rs` for info. Although it was probably caused by me in this PR since I hacked the cache to support zero hashes as per the BIP (previously we were only serializing zeros and not setting the cache fields).

Anyways, open to any ideas if you have time to look at this.

Thanks

https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki

sanket1729 commented on bitcoin/src/blockdata/script.rs:
```diff
@@ -1184,16 +1184,21 @@ impl ScriptBuf {
         let script = Builder::new()
-            .push_opcode(OP_DUP)
-            .push_opcode(OP_HASH160)
+            // Why don't we push OP_PUSHBYTES_25 (0x19)
```
These are two bugs making canceling out each other and making things work. In all cases except code separator(that we explicitly don't support). `script_code` is just serialized `Script` (spending script pubkey to be precise). When serializing `Script`, we prefix it with its length and that is where the 019 (25) comes from. 

According to the BIP, the script_code should also prefix and the sighash function should accept the `script_code` as input. The correct way to fix this would be to change the `segwit_signature_hash` API. 

One suggestion, feel free to improvise:


```rust

pub struct Bip143ScriptCode(Script);

impl Bip143ScriptCode {
    pub fn new_wpkh(wpkh: Script) -> Option<Self>;
    pub fn new_wsh(wsh: Script) -> Option<Self>;
    // Assume the given script as the correctly constructed script code. Useful for custom cases with code separator, or in cases where the user wants to explicitly trust the given script as correctly computed script code.
    pub fn dangerous_assume_script_code(raw_script: Script) -> Self;
}

// New API that takes script code as BIP143ScriptCode instead of raw &Script.
// Deprecate the previous `segwit_signature_hash` API. I wanted a reason to add the `v0` there :P 
pub fn segwit_v0_signature_hash( .. ,script_code: &Bip143ScriptCode, ...)
```

tcharding replied:
> When serializing Script, we prefix it with its length

Aaaah thanks man, I missed that. For any other readers
```rust
impl Encodable for Script {
    #[inline]
    fn consensus_encode<W: io::Write + ?Sized>(&self, w: &mut W) -> Result<usize, io::Error> {
        crate::consensus::encode::consensus_encode_with_size(&self.0, w)
    }
}
```
and 
```rust
pub(crate) fn consensus_encode_with_size<S: io::Write>(data: &[u8], mut s: S) -> Result<usize, io::Error> {
    let vi_len = VarInt(data.len() as u64).consensus_encode(&mut s)?;
    s.emit_slice(data)?;
    Ok(vi_len + data.len())
}
```

Will hack up something based on your suggestion. Thanks again.

Kixunil replied:
Isn't this related to the tagged scripts? Wouldn't it be better to implement script tagging first?

tcharding replied:
What is "script tagging"? I'm unable to find any reference to it on the internet.

apoelstra replied:
@tcharding parameterizing the `Script` type by a dummy type indicating its purpose (scriptSig, scriptPubKey, witnessScript, taproot/segwit/legacy, etc)

tcharding replied:
ahhh, nice one, I was wondering why we didn't express the script purpose anywhere.

tcharding replied:
Currently your suggestion is implemented @sanket1729 but I'm going to have a crack at script tagging.

sanket1729 reviewed (commented):
Left a detailed comment with a suggestion.

tcharding commented:
I've added a new type `bip143::ScriptCode` as suggested. I've also added a test for the P2SH-P2WPKH test vector which led me to add a couple of methods to `ScriptBuf` (after much investigating), please let me know if I'm off track.

I also noticed that unit tests in `sighash` already use the test vectors from BIP143, so the internal sighash cache stuff is tested already, I missed that before.

I'll continue with the other test vectors.

Kixunil commented on bitcoin/src/bip143.rs:
```diff
@@ -0,0 +1,168 @@
+
+    /// Returns the script code used for spending a P2WSH output using `witness` (redeem script).
+    ///
+    /// WARNING: Does not remove `OP_CODESEPARATOR`.
```
IIUC if you supply a script with `OP_CODESEPARATOR` this will create a money-losing script. If it's true, I want the function to panic rather than losing money.

tcharding replied:
Good point, thanks man.

tcharding replied:
Alright, I might sound like a noob here, please correct me if I'm wrong, but this function is for determining the scriptCode which is used to unlock coins, it has nothing to do with the locking script so it cannot create a money-losing script.

Its not trivial to determine if the witness has `OP_CODESEPARATOR` in it because one has to have knowledge of the script structure since `0xab` can appear inside other data in the witness.

Kixunil replied:
Oh, I think you're right, TBH my mind wasn't perfectly clear when I wrote that. I wonder if we should protect against creation of the address instead (where we can, maybe as a debug assert).

Regarding implementation shouldn't we just iterate over `Instructions`?

sanket1729 replied:
`OP_CODESEP` is just an esoteric thing that users must know what they are doing. I think writing in doc-comment is just fine. The engineering time we waste here is not going to be worth any speculated use-cases (if any) that we can think. 

I vote that we can keep the codesep issue as a very low-priority until we have someone complain about it.

sanket1729 replied:
Some other useful things we can do here is to check the witness script size < 3600 (standardness) rules and < 10000 (consensus) rule.

Kixunil replied:
If calling `instructions().find(|x| matches!(x, Instrution::Op(OP_CODESEPARATOR)))` is enough, then I really want it - low effort protection against inconvenience (since *our library* doesn't support it). If it's more than that, I'd probably ignore it.

sanket1729 commented on bitcoin/src/bip143.rs:
```diff
@@ -0,0 +1,175 @@
+impl ScriptCode {
+    /// Returns the script code used for spending a P2WPKH output if this script is a script pubkey
+    /// for a P2WPKH output.
+    pub fn new_p2wpkh(wpkh: &Script) -> Option<Self> {
```
This should not be `&Script`. It should either be `WPubKeyHash` or preferably `PublicKey`.

tcharding replied:
Just saw this comment @sanket1729, posting so you know I haven't ignored it. I had a quick go but I'm too brain fried to add it today. I'll make a note though. Thanks

tcharding commented:
Closing in favour of #1543
